### PR TITLE
Fix failed to create EnvoyFilter in other namespaces

### LINKED ISSUE #27165: failed to create "EnvoyFilter/istio-system/stats-filter-1.7": namespaces "istio-system" not found
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**

I was trying to reproduce https://github.com/istio/istio/issues/25498 and got the following error.

```
$ ./out/linux_amd64/istioctl install --set profile=minimal --set values.global.istioNamespace=istio1 --set hub=grc.io/istio-testing  -d manifests/
This will install the Istio profile into the cluster. Proceed? (y/N) y

✔ Istio core installed                                                                                                                                                                            
2020-09-09T10:42:09.832101Z     error   installer       failed to create "EnvoyFilter/istio-system/metadata-exchange-1.6": namespaces "istio-system" not found
2020-09-09T10:42:09.886275Z     error   installer       failed to create "EnvoyFilter/istio-system/metadata-exchange-1.7": namespaces "istio-system" not found
2020-09-09T10:42:09.939994Z     error   installer       failed to create "EnvoyFilter/istio-system/metadata-exchange-1.8": namespaces "istio-system" not found
2020-09-09T10:42:09.994269Z     error   installer       failed to create "EnvoyFilter/istio-system/stats-filter-1.6": namespaces "istio-system" not found
2020-09-09T10:42:10.048425Z     error   installer       failed to create "EnvoyFilter/istio-system/stats-filter-1.7": namespaces "istio-system" not found
2020-09-09T10:42:10.102391Z     error   installer       failed to create "EnvoyFilter/istio-system/stats-filter-1.8": namespaces "istio-system" not found
2020-09-09T10:42:10.156077Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.6": namespaces "istio-system" not found
2020-09-09T10:42:10.209545Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.7": namespaces "istio-system" not found
2020-09-09T10:42:10.262965Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.8": namespaces "istio-system" not found
2020-09-09T10:42:10.316949Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.6": namespaces "istio-system" not found
2020-09-09T10:42:10.371008Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.7": namespaces "istio-system" not found
2020-09-09T10:42:10.424881Z     error   installer       failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.8": namespaces "istio-system" not found
✘ Istiod encountered an error: failed to create "EnvoyFilter/istio-system/metadata-exchange-1.6": namespaces "istio-system" not found                                                             
failed to create "EnvoyFilter/istio-system/metadata-exchange-1.7": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/metadata-exchange-1.8": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/stats-filter-1.6": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/stats-filter-1.7": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/stats-filter-1.8": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.6": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.7": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-metadata-exchange-1.8": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.6": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.7": namespaces "istio-system" not found
failed to create "EnvoyFilter/istio-system/tcp-stats-filter-1.8": namespaces "istio-system" not found
- Pruning removed resources                                                                                                                                                                       Error: failed to install manifests: errors occurred during operation
```

**Affected product area (please put an X in all that apply)**

[ ] Docs
[X] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Expected behavior**

**Steps to reproduce the bug**
```
$ make build
$ ./out/linux_amd64/istioctl install --set profile=minimal --set values.global.istioNamespace=istio1 --set hub=grc.io/istio-testing  -d manifests/
```

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**

```
$ ./out/linux_amd64/istioctl version --remote
no running Istio pods in "istio-system"
1.8-dev

$ kubectl version --short
Client Version: v1.18.2
Server Version: v1.18.2
```

**How was Istio installed?**
Istio was not installed

**Environment where bug was observed (cloud vendor, OS, etc)**
```
kind v0.8.1 go1.13.3 linux/amd64
RHEL-8
```

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

### PR DESCRIPTION
**Pull request by shamsher31**
To fix https://github.com/istio/istio/issues/27165 and hopefully https://github.com/istio/istio/issues/26813

Different ways of setting namespaces during installation https://github.com/istio/istio/issues/23401#issuecomment-621916406.

```
--set values.global.istioNamespace=<custom_ns> 
--set meshConfig.rootNamespace=<custom_ns>
```

Related:
@howardjohn comment https://github.com/istio/istio/issues/23401#issuecomment-621543877
rootNamespace : https://github.com/istio/istio/pull/23407

### CODE REVIEW

File: manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml [hunk 1]
```diff
@@ -6,6 +6,8 @@ metadata:
   name: metadata-exchange-1.7{{- if not (eq .Values.revision "") }}-{{ .Values.revision }}{{- end }}
   {{- if .Values.meshConfig.rootNamespace }}
   namespace: {{ .Values.meshConfig.rootNamespace }}
+  {{- else if .Values.global.istioNamespace }}
+  namespace: {{ .Values.global.istioNamespace }}
```
Comment (morvencao) [inline manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: I don't think this fix the original issue, `{{ .Release.Namespace }}`, looks like the default value for `meshConfig.rootNamespace` is always `istio-system`.

Comment (morvencao) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: See: https://github.com/istio/istio/blob/master/manifests/charts/istio-control/istio-discovery/values.yaml#L166

Comment (shamsher31) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: I couldn't find from where this value `{{ .Release.Namespace }}` is set. Maybe removing default value for `meshConfig.rootNamespace` will work. I will give it a try.

Comment (shamsher31) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: > looks like the default value for meshConfig.rootNamespace is always istio-system

Yes, and it always assumes default value and else does not execute.


> In that case we have to set two extra namespaces vis old helm values:
> ```
> apiVersion: install.istio.io/v1alpha1
> kind: IstioOperator
> metadata:
>   namespace: istio-system-test
>   name: "example"
> spec:
>   namespace: istio-system-test
>   values:
>     global:
>       istioNamespace: istio-system-test
>     meshConfig:
>       rootNamespace: istio-system-test
> ```

@morvencao Does that means both the flags `global.istioNamespace` and `meshConfig.rootNamespace`  needs to be set?

Comment (morvencao) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: Yes.
We may need figure out how `namespace` field in iop spec maps to `istioNamespace` and `rootNamespace` if we donit expect  to set two extra namespaces vis old helm values.

Comment (shamsher31) [reply manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.7.yaml:10]: Thanks, I will look into it.

**Review by howardjohn — COMMENTED** [review]
Can you explain what the 3 namespaces mean? This seems a bit off to me but I am not sure I understand the distinction between istio namespace and release namespace?

**Comment by shamsher31** [comment]
As far as I understood, installing istio in a custom namespace is broken because the default value for `rootNamespace`  is set to `istio-system`. (I removed the default value in this PR)

There is no provision for accepting custom namespace if user pass 
`--set values.global.istioNamespace=<custom_ns> `. So even if you use `values.global.istioNamespace=<custom_ns>`, it breaks because it couldn't find a custom namespace.

And like I said  https://github.com/istio/istio/pull/27200#discussion_r486182141. I assumed that if the user does not pass anything, it uses the release namespace which might be `istio-system`. I may be wrong here.

Also I'm still not convinced by so many ways of passing custom namespaces.

**Comment by howardjohn** [comment]
First, we need to understand the rootNamespace != the installation namespace. This must be preserved. It is usually the same namespace for people that don't care about splitting them, and its a good idea to make the default for rootNamespace dynamic (always matching installation namespace) rather than statically istio-system but it must be overridable if the user sets it explicitly.

As far as Release.Namespace vs istioNamespace - I believe the distinction is that other components may reference each other. For example, the gateway needs to know the istioNamespace (it may make more sense logically to think of it as "istiod" namespace perhaps), so it can connect. However, it still needs .Release.Namespace since ingress can be installed anywhere.

So I think, without looking into it too deeply, it may make sense to have all 3 of these defaulted to the same thing but they must all be independently overwriteable

**Comment by shamsher31** [comment]
Currently, I'm unable to build it locally. Not sure if anything changed with `1.8-dev` images?

```
$ ./out/linux_amd64/istioctl install --set hub=grc.io/istio-testing -d manifests/
```

```
Failed to pull image "grc.io/istio-testing/pilot:1.8-dev"
```

```
Events:
  Type     Reason     Age                    From                                  Message
  ----     ------     ----                   ----                                  -------
  Normal   Scheduled  6m11s                  default-scheduler                     Successfully assigned istio-system/istiod-5ff4b48497-jvkrn to istio-testing-control-plane
  Normal   Pulling    2m58s (x4 over 6m12s)  kubelet, istio-testing-control-plane  Pulling image "grc.io/istio-testing/pilot:1.8-dev"
  Warning  Failed     2m28s (x4 over 5m42s)  kubelet, istio-testing-control-plane  Failed to pull image "grc.io/istio-testing/pilot:1.8-dev": rpc error: code = Unknown desc = failed to pull and unpack image "grc.io/istio-testing/pilot:1.8-dev": failed to resolve reference "grc.io/istio-testing/pilot:1.8-dev": failed to do request: Head https://grc.io/v2/istio-testing/pilot/manifests/1.8-dev: dial tcp 64.98.145.30:443: i/o timeout
  Warning  Failed     2m28s (x4 over 5m42s)  kubelet, istio-testing-control-plane  Error: ErrImagePull
  Warning  Failed     2m14s (x6 over 5m41s)  kubelet, istio-testing-control-plane  Error: ImagePullBackOff
  Normal   BackOff    71s (x11 over 5m41s)   kubelet, istio-testing-control-plane  Back-off pulling image "grc.io/istio-testing/pilot:1.8-dev"
```

**Comment by howardjohn** [comment]
grc.io -> gcr.io should do the trick.

**Comment by shamsher31** [comment]
> grc.io -> gcr.io should do the trick.

Ahh, silly mistake. Thanks for catching it.

**Comment by shamsher31** [comment]
Currently `install` is not able to read namespace passed in `values.global.istioNamespace=istio-test`. If I create `istio-test` namespace manually and run `install` everything works fine.

```
$ ./out/linux_amd64/istioctl install --set profile=minimal --set values.global.istioNamespace=istio-test --set hub=gcr.io/istio-testing  -d manifests/
```

File: manifests/charts/base/templates/validatingwebhookconfiguration.yaml [hunk 2]
```diff
@@ -15,7 +15,9 @@ webhooks:
       {{- else }}
       service:
         name: istiod
+        {{- if .Values.global.istioNamespace }}
```
Comment (howardjohn) [inline manifests/charts/base/templates/validatingwebhookconfiguration.yaml:18]: this field is required.

File: manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.6.yaml [hunk 3]
```diff
@@ -3,7 +3,9 @@ apiVersion: networking.istio.io/v1alpha3
 kind: EnvoyFilter
 metadata:
   name: metadata-exchange-1.6{{- if not (eq .Values.revision "") }}-{{ .Values.revision }}{{- end }}
-  {{- if .Values.meshConfig.rootNamespace }}
+  {{- if .Values.global.istioNamespace }}
+  namespace: {{ .Values.global.istioNamespace }}
+  {{- else if .Values.meshConfig.rootNamespace }}
```
Comment (howardjohn) [inline manifests/charts/istio-control/istio-discovery/templates/telemetryv2_1.6.yaml:8]: root namespace takes priority

**Review by howardjohn — CHANGES_REQUESTED** [review]

File: manifests/charts/base/templates/validatingwebhookconfiguration.yaml (hunk 2 above)
Comment (shamsher31) [reply manifests/charts/base/templates/validatingwebhookconfiguration.yaml:18]: Not sure If I updated it or it gets updated during `make gen`. But I will verify.

**Pull request closed without merging**